buttercup/cogs/heatmap.py

```diff
@@ -72,6 +76,101 @@ def create_file_from_heatmap(
     return File(heatmap_table, "heatmap_table.png")
 
 
+def _get_week_index(date: datetime) -> int:
+    """Get an identification number for the week.
+
+    year * 100 + week
+    """
+    calendar = date.isocalendar()
+    return calendar[0] * 100 + calendar[1]
+
+
+def _get_month_annotations(activity_df: pd.DataFrame) -> List[str]:
+    """Get the month annotations for the activity map.
+
+    The first week that has a date of a new month will get the month
+    as annotation.
+    """
+    annotations = []
+
+    for col in activity_df.columns:
+        # Extract the year and week number from the column name
+        year = col // 100
+        week = col - year * 100
+
+        # Reconstruct all dates in this week
+        dates = [datetime.fromisocalendar(year, week, day) for day in range(1, 8)]
+
+        new_month = ""
+
+        for date in dates:
+            if date.day == 1:
+                # New month, add the month name
+                new_month = date.strftime("%b")
+                break
+
+        annotations.append(new_month)
+
+    return annotations
+
+
+def _create_file_from_activity_map(
+    activity_df: pd.DataFrame, user: Optional[BlossomUser], time_str: str
+) -> File:
+    """Create a Discord file containing the activity map."""
+    days = i18n["heatmap"]["days"]
+
+    # Only annotate the maximum values
+    max_value = activity_df.max().max()
+    annotations = activity_df.apply(
+        lambda series: series.apply(
+            lambda value: f"{value:0.0f}" if value == max_value else ""
+        )
+    )
+
+    fig, ax = plt.subplots()
+    fig: plt.Figure
+    ax: plt.Axes
+    fig.set_size_inches(9, 3.44)
+    fig.subplots_adjust(bottom=0.25, top=1, left=0.05, right=0.98, wspace=0, hspace=0)
+
+    cbar_kws = {
+        "orientation": "horizontal",
+        "fraction": 0.08,
+        "aspect": 40,
+        "shrink": 0.6,
+    }
+
+    sns.heatmap(
+        activity_df,
+        ax=ax,
+        annot=annotations,
+        annot_kws={"fontsize": "x-small"},
+        fmt="s",
+        cbar=True,
+        cbar_kws=cbar_kws,
+        square=True,
+        yticklabels=days,
+        xticklabels=_get_month_annotations(activity_df),
+    )
+
+    ax.set_title(
+        i18n["activity"]["plot_title"].format(
+            user=get_username(user, escape=False), time=time_str
+        )
+    )
+    # Remove axis labels
+    ax.set_xlabel(None)
+    ax.set_ylabel(None)
+
+    activity_map = io.BytesIO()
+    plt.savefig(activity_map, format="png")
+    activity_map.seek(0)
+    plt.close(fig)
+
+    return File(activity_map, "activity_map.png")
+
+

```

Comment:
Random question: When writing this big sections of code for the yearly heatmap, do you have precedence that you follow? or do you kinda freehand it?

Reply:
In this case I could roughly follow the implementation of the `/heatmap` command, as it works in a similar way.
Usually, things start out simpler and I add new stuff to fix bugs or polish a feature.

At some point though you have to do "refactoring", i.e. restructuring the code without changing functionality, such that it becomes easier to read and maintain.